src/control/lib/control/pool.go

```diff
@@ -744,80 +744,183 @@ func PoolGetProp(ctx context.Context, rpcClient UnaryInvoker, req *PoolGetPropRe
+// PoolRanksResp struct contains response from operation on multiple pool-ranks.
+type PoolRanksResp struct {
+	Status       int32           `json:"status"`
+	ID           string          `json:"id"`
+	FailedRanks  []ranklist.Rank `json:"failed_ranks"`
+	SuccessRanks []ranklist.Rank `json:"success_ranks"`
+	InitialRanks []ranklist.Rank `json:"initial_ranks"`
+}
 
-// PoolExclude will set a pool target for a specific rank to down.
-// This should automatically start the rebuildiing process.
-// Returns an error (including any DER code from DAOS).
-func PoolExclude(ctx context.Context, rpcClient UnaryInvoker, req *PoolExcludeReq) error {
-	pbReq := &mgmtpb.PoolExcludeReq{
-		Sys:       req.getSystem(rpcClient),
-		Id:        req.ID,
-		Rank:      req.Rank.Uint32(),
-		TargetIdx: req.TargetIdx,
-		Force:     req.Force,
+// Errors returns either a generic failure based on the requested rankset failure or a rank-specific
+// failure depending on whether or not FailedRank has been set in the response.
+func (resp *PoolRanksResp) Errors() error {
+	if resp == nil {
+		return errors.Errorf("nil %T", resp)
+	}
+	if resp.Status == int32(daos.Success) {
+		return nil
+	}
+
+	err := daos.Status(resp.Status)
+
+	if len(resp.FailedRanks) > 0 {
+		rs := ranklist.RankSetFromRanks(resp.InitialRanks)
+		return errors.Wrapf(err, "pool %s ranks %s", resp.ID, rs.String())
+	}
+
+	rs := ranklist.RankSetFromRanks(resp.FailedRanks)
+	return errors.Wrapf(err, "pool %s ranks %s", resp.ID, rs.String())
+}
+
+// GetResults returns a slice of results from the response and input error.
+func (resp *PoolRanksResp) GetResults(poolID string, reqRanks []ranklist.Rank, errIn error) (results []*PoolRanksResult, err error) {
+	if poolID == "" {
+		return nil, errors.New("pool id is missing")
+	}
+
+	defer func() {
+		err = errors.Wrapf(err, "pool %s", resp.ID)
+	}()
+
+	if errIn != nil {
+		if len(reqRanks) == 0 {
+			return nil, errors.New("reqRanks is missing")
+		}
+
+		// Return root cause so rank results can be aggregated if required.
+		msgErr := errIn.Error()
+		if f, ok := errors.Cause(errIn).(*fault.Fault); ok {
+			msgErr = f.Error()
+		}
+		results = append(results, &PoolRanksResult{
+			ID:     poolID,
+			Ranks:  reqRanks,
+			Status: int32(daos.MiscError),
+			Msg:    msgErr,
+		})
+
+		return results, nil
 	}
+
+	if resp == nil {
+		return nil, errors.Errorf("nil %T", resp)
+	}
+
+	if resp.Status != int32(daos.Success) {
+		if len(resp.FailedRanks) == 0 {
+			return nil, errors.New("no failed rank returned with non-zero status")
+		}
+		// Add one result for failed ranks. Note same status applies to all failed ranks.
+		results = append(results, &PoolRanksResult{
+			ID:     poolID,
+			Ranks:  resp.FailedRanks,
+			Status: resp.Status,
+			Msg:    daos.Status(resp.Status).Error(),
+		})
+	} else {
+		if len(resp.FailedRanks) > 0 {
+			return nil, errors.New("failed rank returned with zero status")
```

Comment:
IMO we should handle these seemingly-inconsistent responses more gracefully. It's a surprising output but seems more like something we should log as strange (maybe NOTICE level?) and return an appropriate result anyway. An error like this isn't really actionable for the user, since it indicates an engine bug in writing the response. The operation may have succeeded or failed, and the lists of successful and failed ranks seem potentially still useful.

Reply:
will be addressed in refactor